### ISSUE TITLE
[Question] Is it possible to access variables in callbacks?

### ISSUE THREAD

**Issue by DebkanchanSamadder** [rekab-app/background_locator#105]
I have been trying to implement this plugin in my app and so far it's very helpful. But I'm trying to make a geofence solution with it and that's where the problems arise.

So far it looks like both `initCallback` and `callback` can access other static variables. But after background locator is registered it can't read the variable's updated value.

example: 
```
static int number = 0
static string uid
...
static _initCallback(Map<dynamic, dynamic> params) async {
   _uid = params["uid"];
  print(number) //will always print 0
 }

static _callback(LocationDto locationDto) {
  final SendPort sendPort = IsolateNameServer.lookupPortByName(_isolateName);
  sendPort?.send(locationDto);
  print(number); //will also always print 0
}

BackgroundLocator.registerLocationUpdate(
  _callback,
  initCallback: _initCallback,
  initDataCallback: {"uid": _user.id},
  disposeCallback: _disposeCallback,
  settings: settings.LocationSettings(
      accuracy: settings.LocationAccuracy.HIGH,
      notificationTitle: "Ride Partner",
      notificationMsg: "Running in Background. Searching for orders",
      wakeLockTime: 1,
      autoStop: false,
      distanceFilter: 0.1,
     interval: 5
  ),
);

number = 1;
```

Also, like how `uid` is initiated via initDataCallback, every time I need to register a variable do I need to unregister the service then reregister again or is there a more elegant way of registering a variable?

**Comment by gpibarra** [comment]
I think you should use shared preferences

**Comment by DebkanchanSamadder** [comment]
> I think you should use shared preferences

is it a good practice to call `SharedPreferences.getInstance()` every time callback fires?

**Comment by gpibarra** [comment]
Which version are you using?
Have you tried doing it as the example says ("location_callback_handler.dart" and "location_service_repository.dart")?

**Comment by DebkanchanSamadder** [comment]
Are you referring to the setting variable from initCallback method? if yes, then you can see it in my code that I have tried it

**Comment by gpibarra** [comment]
No. in the example use a whole container class (facade) to execute the callbacks

**Comment by DebkanchanSamadder** [comment]
Tried that but the problem remains. The variables are okay as long as the main thread is on. But when it's restarted everything goes back to null. Looks like the only option is to write to a file and read from it which the example app does too for saving the location logs. But isn't calling `await getTemporaryDirectory()` expensive?

**Comment by DebkanchanSamadder** [comment]
Update: Tested again with repository method and it actually works. The problem isn't with any of this. I have observed a strange behaviour. If you open and close the main app it's ok BUT if you **hot restart** once or twice it goes back to null

**Comment by gpibarra** [comment]
What device are you using? you might have a power / battery gestures that limit the use of apps in the background (i think huawei has that pre-installed)

**Comment by DebkanchanSamadder** [comment]
Tested on both Asus Zenfone Max Pro M1 (X00TD) and OnePlus 8 and both time hot restarting once or twice wipes the variables of the repository

**Comment by DebkanchanSamadder** [comment]
Right now I'm solving this issue by not enabling restart on boot. Using the repository method to store variables and callbacks and also using local files to save data that needs to be persisted across boot.

**Comment by DebkanchanSamadder** [comment]
I think currently this is the only solution. Closing issue.

**Comment by mehdok** [comment]
Right now I think the best solution is to store you variable into `sharePreferences`. please take a look at [#91](https://github.com/rekab-app/background_locator/issues/92#event-3611486960) If you decided to use `SharedPreferences` plugin.

**Comment by DebkanchanSamadder** [comment]
@mehdok  Thank you for the pointer. I didn't know you cloud reload everything from filesystem to memory with shared preferences. My custom solution is in theory like that but there's also managing files and saving the location of those etc. In contrast this'll be much easier.

Although I have one question. Can you use the repository to save `SharedPreferences` instance from `await SharedPreferences.getInstance()` in initcallback as a variable so you don't have to call `getInstance()` on every callback?

**Comment by mehdok** [comment]
@DebkanchanSamadder Honestly I don't know if it is possible:) The `initcallback` is just for primitive types, but You can test that. But I use the `sharedPreferences` this way and each time I call `SharedPreferences.getInstance()`, I think the overhead of calling this is very little;

**Comment by DebkanchanSamadder** [comment]
Thanks @mehdok! I refrained from setting any variables from inside location service repository because it doesn't play nicely with hot restart, basically it wipes those variables. This issue might be more deep rooted so I avoided it totally. Using `SharedPreferences.getInstance()` and `SharedPreferences.reload()` worked like charm. In fact it worked so good I think it should be the official way of sharing data and should be in documentation.
